fragment.go

```diff
@@ -1454,65 +1503,109 @@ func (f *fragment) bulkImport(rowIDs, columnIDs []uint64, options *ImportOptions
 	return f.bulkImportStandard(rowIDs, columnIDs, options)
 }
 
-// bulkImportStandard performs a bulk import on a standard fragment.
-func (f *fragment) bulkImportStandard(rowIDs, columnIDs []uint64, options *ImportOptions) error {
-	// Create a temporary bitmap which will be populated by rowIDs and columnIDs
-	// and then merged into the existing fragment's bitmap.
-	localBitmap := roaring.NewBitmap()
-
-	// Disconnect op writer so we don't append updates.
-	localBitmap.OpWriter = nil
-
-	// rowSet maintains the set of rowIDs present in this import.
-	// It allows the cache to be updated once per row, instead of once
-	// per bit.
+// bulkImportStandard performs a bulk import on a standard fragment. May mutate
+// its rowIDs and columnIDs arguments.
+func (f *fragment) bulkImportStandard(rowIDs, columnIDs []uint64, options *ImportOptions) (err error) {
+	// rowSet maintains the set of rowIDs present in this import. It allows the
+	// cache to be updated once per row, instead of once per bit. TODO: consider
+	// sorting by rowID/columnID first and avoiding the map allocation here. (we
+	// could reuse rowIDs to store the list of unique row IDs)
 	rowSet := make(map[uint64]struct{})
 	lastRowID := uint64(0)
 
-	// Process every bit by writing to a local bitmap,
-	// to be merged with fragment storage next.
-	for i := range rowIDs {
+	// replace columnIDs with calculated positions to avoid allocation.
+	for i := 0; i < len(columnIDs); i++ {
 		rowID, columnID := rowIDs[i], columnIDs[i]
-
-		// Determine the position of the bit in the storage.
 		pos, err := f.pos(rowID, columnID)
 		if err != nil {
 			return err
 		}
-
-		// Write to local storage.
-		_, err = localBitmap.Add(pos)
-		if err != nil {
-			return err
-		}
-
-		// Reduce the StatsD rate for high volume stats
-		f.stats.Count("ImportBit", 1, 0.0001)
+		columnIDs[i] = pos
 
 		// Add row to rowSet.
 		if i == 0 || rowID != lastRowID {
 			lastRowID = rowID
 			rowSet[rowID] = struct{}{}
 		}
 	}
-
+	positions := columnIDs
 	f.mu.Lock()
 	defer f.mu.Unlock()
+	if options.Clear {
+		err = f.importPositions(nil, positions, rowSet)
+	} else {
+		err = f.importPositions(positions, nil, rowSet)
+	}
+	return errors.Wrap(err, "bulkImportStandard")
+}
+
+// importPositions takes slices of positions within the fragment to set and
+// clear in storage. One must also pass in the set of unique rows which are
+// affected by the set and clear operations. It is unprotected (f.mu must be
+// locked when calling it). No position should appear in both set and clear.
+//
+// importPositions tries to intelligently decide whether or not to do a full
+// snapshot of the fragment or just do in-memory updates while appending
+// operations to the op log.
+func (f *fragment) importPositions(set, clear []uint64, rowSet map[uint64]struct{}) error {
```

Comment:
reading this through again, i feel like these two paths are different enough that it might be cleaner/simpler to have them be two separate code flows. going back and forth, realizing that in smallWrite all of setBitmap, clearBitmap, and results end up being f.storage, and re-reading the logic to see why we sometimes have clearBitmap.AddN and sometimes RemoveN, seems complex. that or it's late in the day and i'm slow now.

Reply:
nah, you're totally right, it's a mess. I'll see about breaking it apart.

Reply:
much cleaner now :)